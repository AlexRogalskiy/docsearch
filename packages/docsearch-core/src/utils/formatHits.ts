import { DocSearchHit, DocSearchHits, AlgoliaHit } from 'docsearch.js-types';

import { getHighlightedValue, getSnippetedValue } from '../utils';

export type AlgoliaHitWithRootLevels = Omit<AlgoliaHit, 'hierarchy'> & {
  lvl0?: AlgoliaHit['_highlightResult']['content'];
  lvl1?: AlgoliaHit['_highlightResult']['content'];
  lvl2?: AlgoliaHit['_highlightResult']['content'];
  lvl3?: AlgoliaHit['_highlightResult']['content'];
  lvl4?: AlgoliaHit['_highlightResult']['content'];
  lvl5?: AlgoliaHit['_highlightResult']['content'];
  lvl6?: AlgoliaHit['_highlightResult']['content'];
};

function groupBy<TValue = any>(
  values: TValue[],
  predicate: (value: TValue) => any
) {
  return values.reduce(function(obj, item) {
    const key = predicate(item);

    if (!obj.hasOwnProperty(key)) {
      obj[key] = [];
    }

    obj[key].push(item);

    return obj;
  }, {});
}

export function copyHierarchyValuesToRoot(
  rawHit: AlgoliaHit
): AlgoliaHitWithRootLevels {
  const levels = Object.entries(rawHit._highlightResult.hierarchy).reduce(
    (acc, [key, level]) => ({ ...acc, [key]: level.value }),
    {}
  );
  const { hierarchy, ...hit } = rawHit;

  return {
    ...hit,
    ...levels,
  };
}

function getUrl(hit: AlgoliaHitWithRootLevels): string {
  const { url, anchor } = hit;

  if (url) {
    const containsAnchor = url.indexOf('#') !== -1;

    if (containsAnchor) {
      return url;
    } else if (anchor) {
      return `${hit.url}#${hit.anchor}`;
    }

    return url;
  } else if (anchor) {
    return `#${hit.anchor}`;
  }

  return '';
}

export function formatHit(hit: AlgoliaHitWithRootLevels): DocSearchHit {
  const levels = [
    getHighlightedValue(hit, 'lvl0'),
    getHighlightedValue(hit, 'lvl1'),
    getHighlightedValue(hit, 'lvl2'),
    getHighlightedValue(hit, 'lvl3'),
    getHighlightedValue(hit, 'lvl4'),
    getHighlightedValue(hit, 'lvl5'),
    getHighlightedValue(hit, 'lvl6'),
  ].filter(Boolean);

  return {
    objectID: hit.objectID,
    levels,
    levelIndex: levels.length - 1,
    content: getSnippetedValue(hit, 'content'),
    url: getUrl(hit),
  };
}

function formatHits(rawHits: AlgoliaHit[]): DocSearchHits {
  const hits = rawHits.map(copyHierarchyValuesToRoot).map(formatHit);
  const formattedHits = groupBy(hits, hit => hit.levels[hit.levelIndex]);

  return formattedHits;
}

export default formatHits;
