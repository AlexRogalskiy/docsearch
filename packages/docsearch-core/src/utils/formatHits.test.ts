import formatHits from './formatHits';
import { docsearchHits } from './fixtures/';

describe('formatHits', () => {
  test('with empty hits', () => {
    const hits = [];
    const formattedHits = formatHits(hits);

    expect(formattedHits).toEqual({});
  });

  test('with actual hits', () => {
    const hits = docsearchHits;
    const formattedHits = formatHits(hits);

    expect(formattedHits).toMatchSnapshot();
  });
});
