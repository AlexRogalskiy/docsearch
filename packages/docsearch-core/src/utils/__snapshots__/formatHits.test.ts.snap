// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatHits with actual hits 1`] = `
Object {
  "Dropdown Behavior": Array [
    Object {
      "content": "…dropdown.  <span class=\\"algolia-docsearch-suggestion--highlight\\">hit</span>sPerPage  set the number of shown <span class=\\"algolia-docsearch-suggestion--highlight\\">hit</span>s …",
      "levelIndex": 2,
      "levels": Array [
        "Adding DocSearch To Your UI",
        "Dropdown Behavior",
        "algoliaOptions",
      ],
      "objectID": "2356570030",
      "url": "https://community.algolia.com/docsearch/behavior.html#algoliaoptions",
    },
    Object {
      "content": "…anchor). You can override results (<span class=\\"algolia-docsearch-suggestion--highlight\\">hit</span>) to
add your own…",
      "levelIndex": 2,
      "levels": Array [
        "Adding DocSearch To Your UI",
        "Dropdown Behavior",
        "handleSelected",
      ],
      "objectID": "2356569750",
      "url": "https://community.algolia.com/docsearch/behavior.html#handleselected",
    },
  ],
  "FAQ": Array [
    Object {
      "content": "The  property  nb_<span class=\\"algolia-docsearch-suggestion--highlight\\">hit</span>s  in your configuration keeps track…",
      "levelIndex": 2,
      "levels": Array [
        "Requirements, Tips, FAQ",
        "FAQ",
        "How many records are created by DocSearch?",
      ],
      "objectID": "2356575660",
      "url": "https://community.algolia.com/docsearch/faq.html#how-many-records-are-created-by-docsearch",
    },
  ],
  "Tips for a good search": Array [
    Object {
      "content": null,
      "levelIndex": 2,
      "levels": Array [
        "Requirements, Tips, FAQ",
        "Tips for a good search",
        "Structure the <span class=\\"algolia-docsearch-suggestion--highlight\\">hie</span>rarchy of information",
      ],
      "objectID": "2356576240",
      "url": "https://community.algolia.com/docsearch/tips.html#structure-the-hierarchy-of-information",
    },
  ],
  "selectors": Array [
    Object {
      "content": null,
      "levelIndex": 3,
      "levels": Array [
        "Let’s Get Technical",
        "Config Files",
        "selectors",
        "nb_<span class=\\"algolia-docsearch-suggestion--highlight\\">hit</span>s   Special",
      ],
      "objectID": "2356572690",
      "url": "https://community.algolia.com/docsearch/config-file.html#nb_hits-special",
    },
  ],
}
`;
